src/client/debugger/jupyter/kernelDebugAdapter.ts

```diff
@@ -0,0 +1,331 @@
+        // we do this in case the kernel is stopped when we attach
+        // This might happen if VS Code or the extension host crashes
+        if (message.type === 'request' && (message as DebugProtocol.Request).command === 'attach') {
+            await this.debugInfo();
+        }
+
+        // after disconnecting, hide the breakpoint margin
+        if (message.type === 'request' && (message as DebugProtocol.Request).command === 'disconnect') {
+            void vscode.commands.executeCommand('notebook.toggleBreakpointMargin', this.notebookDocument);
+        }
+
+        // map Source paths from VS Code to Ipykernel temp files
+        this.getMessageSourceAndHookIt(message, (source) => {
+            if (source && source.path) {
+                const p = this.cellToFile.get(source.path);
+                if (p) {
+                    source.path = p;
+                }
+            }
+        });
+
+        if (message.type === 'request') {
+            const request = debugRequest(message as DebugProtocol.Request);
+            const control = this.jupyterSession.requestDebug({
+                seq: request.content.seq,
+                type: 'request',
+                command: request.content.command,
+                arguments: request.content.arguments
+            });
+
+            if (control) {
+                control.onReply = (msg) => {
+                    this.getMessageSourceAndHookIt(msg.content, (source) => {
+                        if (source && source.path) {
+                            const cell = this.fileToCell.get(source.path);
+                            if (cell) {
+                                source.name = path.basename(cell.document.uri.path);
+                                if (cell.index >= 0) {
+                                    source.name += `, Cell ${cell.index + 1}`;
+                                }
+                                source.path = cell.document.uri.toString();
+                            }
+                        }
+                    });
+
+                    this.sendMessage.fire(msg.content);
+                };
+                control.onIOPub = (msg) => {
+                    this.getMessageSourceAndHookIt(msg.content as DebugProtocol.ProtocolMessage, (source) => {
+                        if (source && source.path) {
+                            const cell = this.fileToCell.get(source.path);
+                            if (cell) {
+                                source.name = path.basename(cell.document.uri.path);
+                                if (cell.index >= 0) {
+                                    source.name += `, Cell ${cell.index + 1}`;
+                                }
+                                source.path = cell.document.uri.toString();
+                            }
+                        }
+                    });
+
+                    this.sendMessage.fire(msg.content);
+                };
+                control.onStdin = (msg) => {
```

Comment:
Does anything actually come in on stdin?

Reply:
yeah, nothing, I'll remove it